source/MaterialXCore/Node.cpp

```diff
@@ -728,6 +728,33 @@ InterfaceElementPtr NodeGraph::getImplementation() const
     return nodedef ? nodedef->getImplementation() : InterfaceElementPtr();
 }
 
+vector<PortElementPtr> NodeGraph::getDownstreamPorts() const
+{
+    vector<PortElementPtr> downstreamPorts;
+    for (PortElementPtr port : getDocument()->getMatchingPorts(getQualifiedName(getName())))
+    {
+        ConstGraphElementPtr graph = nullptr;
+        ElementPtr node = port->getParent();
+        if (node)
+        {
+            node = node->getParent();
+            if (node->isA<GraphElement>())
+            {
+                graph = node->asA<GraphElement>();
+            }
+        }
+        if (graph && graph->getChild(getName()) == getSelf())
+        {
+            downstreamPorts.push_back(port);
+        }
```

Comment:
This looks to be the same. I think it should still test if the parent is indeed a GraphElement as the code does this and checks for `node` child and not just "a child"

Reply:
This should work to any depth, so if / when nested nodegraphs are supported it should still work.

Reply:
Just focusing on the current syntax in 1.38 for now, is the restriction to ports on nodes at the root scope of the document strictly correct, or are there other locations in which downstream ports might be found?  For example, would this new function handle examples like the following?

https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/resources/Materials/TestSuite/stdlib/nodegraph_inputs/cascade_nodegraphs.mtlx

If the restriction looks correct to you, then I'd recommend simplifying this logic as far as possible, perhaps using the proposal above as a starting point.  That should help future readers to better understand the cases that we're supporting and not supporting.

Reply:
I tried this out and it does work for this type of configuration. 
Here's a simplified version:
```
<?xml version="1.0"?>
<materialx version="1.38">
  <nodegraph name="nodegraph1">
    <input name="input_color3" type="color3" xpos="1.188406" ypos="-0.810345" output="output_color3" nodegraph="nodegraph2" />
    <surface_unlit name="surface_unlit" type="surfaceshader" xpos="3.246377" ypos="-1.232759">
      <input name="emission_color" type="color3" interfacename="input_color3" />
    </surface_unlit>
    <surfacematerial name="surfacematerial" type="material" xpos="5.884058" ypos="-2.060345">
      <input name="surfaceshader" type="surfaceshader" nodename="surface_unlit" />
    </surfacematerial>
  </nodegraph>
  <nodegraph name="nodegraph2">
    <output name="output_color3" type="color3" nodename="add_color3" xpos="4.992754" ypos="-1.344828" />
    <add name="add_color3" type="color3" xpos="3.246377" ypos="-1.482759">
      <input name="in2" type="color3" value="0.792176, 0.307071, 0.0600427" />
      <input name="in1" type="color3" value="0.0860289, 0.342787, 0.858191" />
    </add>
  </nodegraph>
</materialx>
```
A change on `add_color3` will go to the `output` up to the document and then to the `input_color3` on the other nodegraph (since it's in the downstream connection cache now) and onwards.

I will simplify the logic.

Reply:
@jstone-lucasfilm 
I've added in new code and retested. Everything looks works as before. Thanks for the simplification.